subsys/mgmt/mcumgr/smp_bt.c

```diff
@@ -89,63 +112,51 @@ static struct conn_param_data *get_conn_param_data(const struct bt_conn *conn)
 }
 
 /* Sets connection parameters for a given conn. */
-static void set_conn_latency(struct bt_conn *conn, bool low_latency)
+static void conn_param_set(struct bt_conn *conn, struct bt_le_conn_param *param)
 {
-	struct bt_le_conn_param params;
-	struct conn_param_data *cpd;
-	struct bt_conn_info info;
 	int ret = 0;
+	struct conn_param_data *cpd = conn_param_data_get(conn);
 
-	cpd = get_conn_param_data(conn);
-
-	ret = bt_conn_get_info(conn, &info);
-	__ASSERT_NO_MSG(!ret);
-
-	if ((low_latency && (info.le.latency == 0)) ||
-	    ((!low_latency) && (info.le.latency == DEFAULT_LATENCY))) {
-		/* Already updated. */
-		return;
-	}
-
-	params.interval_min = info.le.interval;
-	params.interval_max = info.le.interval;
-	params.latency = (low_latency) ? (0) : (DEFAULT_LATENCY);
-	params.timeout = info.le.timeout;
-
-	ret = bt_conn_le_param_update(cpd->conn, &params);
+	ret = bt_conn_le_param_update(conn, param);
 	if (ret && (ret != -EALREADY)) {
-		if (!low_latency) {
-			/* Try again to avoid stucking in low latency. */
-			(void)k_work_reschedule(&cpd->dwork, K_MSEC(RESTORE_RETRY_TIME));
-		}
+		/* Try again to avoid being stuck with incorrect connection parameters. */
+		(void)k_work_reschedule(&cpd->ework, K_MSEC(RETRY_TIME));
+	} else {
+		(void)k_work_cancel_delayable(&cpd->ework);
 	}
 }
 
 
-/* Work handler function for restoring the default latency for the connection. */
-static void restore_default_latency(struct k_work *work)
+/* Work handler function for restoring the preferred connection parameters for the connection. */
+static void conn_param_on_pref_restore(struct k_work *work)
 {
-	struct conn_param_data *cpd;
+	struct conn_param_data *cpd = CONTAINER_OF(work, struct conn_param_data, dwork);
 
-	cpd = CONTAINER_OF(work, struct conn_param_data, dwork);
+	conn_param_set(cpd->conn, CONN_PARAM_PREF);
```

Comment:
Not so sure I like the idea of this, if a device connects, and doesn't like the default connection parameters, it will then pick something else, in that case this is going to try to restore it back to the defaults that didn't work. The one that comes to mind would be trying to set a peripheral latency of 3 or 4 on an iphone, from memory (it's been a few years) if you try it a couple of times, the phone will just kill the connection

Reply:
This is a request from the peripheral which can be totally ignored by the master. It is used to change the SMP-optimized parameters (requested during the activity on the SMP characteristic) to the peripheral preferred connection parameters (when there is no activity for 5s on the SMP characteristic). This is request is attempted only once after activity on the SMP characteristic disappears. Do you still think this implementation will be incompatible with iOS devices?

The same request is sent by the Bluetooth stack after 5s from the time when a connection with a master device is established.